internal/pkg/log/zap/zap.go

```diff
@@ -0,0 +1,128 @@
+// Package zap sets up a logger using the Zap logging framework.
+package zap
+
+import (
+	"io"
+	"os"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Options contains all possible log settings.
+type Options struct {
+	// CallerSkip
+	CallerSkip int
+	// Level configures the log verbosity. Defaults to Debug.
+	Level *zap.AtomicLevel
+	// StacktraceLevel is the level which stacktraces will be emitted. Defaults
+	// to Warn.
+	StacktraceLevel *zap.AtomicLevel
+	// Encoder configures how Zap will encode the output. Defaults to JSON.
+	Encoder zapcore.Encoder
+	// OutputDest controls the destination of the log output. Defaults to
+	// os.Stderr.
+	OutputDest io.Writer
+	// ZapOptions allows passing additional optional zap.Options, ex: Sampling.
+	ZapOptions []zap.Option
+}
+
+// Opts allows manipulation of the Zap options.
+type Opts func(*Options)
+
+// addDefaults adds defaults to the Options
+func (o *Options) addDefaults() {
+	if o.CallerSkip < 1 {
+		o.CallerSkip = 1
+	}
+	if o.OutputDest == nil {
+		o.OutputDest = os.Stderr
+	}
+	if o.Encoder == nil {
+		encCfg := zap.NewProductionEncoderConfig()
+		o.Encoder = zapcore.NewJSONEncoder(encCfg)
+	}
+	if o.Level == nil {
+		level := zap.NewAtomicLevelAt(zap.DebugLevel)
+		o.Level = &level
+	}
+	if o.StacktraceLevel == nil {
+		level := zap.NewAtomicLevelAt(zap.WarnLevel)
+		o.StacktraceLevel = &level
+	}
+
+	o.ZapOptions = append(o.ZapOptions, zap.AddStacktrace(o.StacktraceLevel))
+}
+
+// New returns a new zap.Logger configured with the passed Options or their
+// defaults.
+func New(opts ...Opts) *zap.Logger {
```

Comment:
is the intention here to isolate the zap logger such that it can be replaced by another logger?

Reply:
Yeah the zap logger is a 3rd party logger and in general I think it's best practice to isolate it from being called throughout the application code base, especially as our codebase grows and we're using the logger everywhere. Having the interface makes it easier to swap out the underlying 3rd party implementation.